### PR TITLE
JIT: Fold some functions with known equal VNs

### PR DESCRIPTION
**Pull request by jakobbotsch**
Seems like pretty low-hanging fruit unless I'm missing something obvious...

### CODE REVIEW

File: src/coreclr/jit/valuenum.cpp [hunk 1]
```diff
@@ -2134,6 +2134,16 @@ ValueNum ValueNumStore::VNForFunc(var_types typ, VNFunc func, ValueNum arg0VN, V
...
+    if ((genTreeOps(func) == GT_EQ) && (arg0VN == arg1VN))
+    {
+        return VNForIntCon(1);
+    }
+
+    if ((genTreeOps(func) == GT_NE) && (arg0VN == arg1VN))
+    {
+        return VNForIntCon(0);
+    }
+
```
Comment (EgorBo) [inline src/coreclr/jit/valuenum.cpp:2146]: I assume we can do more if it catches diffs e.g. OR(vn0, vn1) -> vn0 for vn0 == vn1 etc 🙂

Comment (jakobbotsch) [reply src/coreclr/jit/valuenum.cpp:2146]: Yep.

Also I need to make sure this does not fold `float.NaN == float.NaN`...

File: src/coreclr/jit/valuenum.cpp [hunk 2]
```diff
@@ -2134,6 +2134,35 @@ ValueNum ValueNumStore::VNForFunc(var_types typ, VNFunc func, ValueNum arg0VN, V
         return EvalFuncForConstantArgs(typ, func, arg0VN, arg1VN);
     }
 
+    if ((arg0VN == arg1VN) && !varTypeIsFloating(TypeOfVN(arg0VN)))
```
Comment (SingleAccretion) [inline src/coreclr/jit/valuenum.cpp:2137]: I would expect such logic to live in `EvalUsingMathIdentity`.

I think it already does (some of) this.

Comment (jakobbotsch) [reply src/coreclr/jit/valuenum.cpp:2137]: Aha, yep, majority of the improvements I was getting was from the wrong float folding :)
Looks like `EvalUsingMathIdentity` does the most important parts of this and the extra identities only result in a small number of improvements in non-test collections, so doesn't seem worth doing.

Comment (EgorBo) [reply src/coreclr/jit/valuenum.cpp:2137]: Still I am pretty sure there are optimizations in morph which we can do here as well 🙂

Comment (jakobbotsch) [reply src/coreclr/jit/valuenum.cpp:2137]:
```
cat morph.cpp >> valuenum.cpp
```

**Pull request closed without merging**